### PR TITLE
OOP exercise

### PR DESCRIPTION
**Pull request by kotlyar-andrey**
### OOP exercise

[Demo](https://kotlyar-andrey.github.io/a-tiny-JS-world/) |
[Code base](https://github.com/kotlyar-andrey/a-tiny-JS-world)

The code is submitted in a dedicated feature branch.

Only code files are submitted.

Please, review.

### CODE REVIEW

File: submissions/kotlyar-andrey/oop_exercise/index.js [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+      this._properties
+        .map((prop) => this[prop])
+        .join("; ")
+        .concat(`; ${friends}`)
+    );
+  }
+}
+
+class HomoSapiens extends Creature {
+  constructor(species, name, gender, phrase, legs, hands) {
```
Comment (OleksiyRudenko) [inline submissions/kotlyar-andrey/oop_exercise/index.js:52]: Do we really want users to specify species for class Homo Sapiens?

File: submissions/kotlyar-andrey/oop_exercise/index.js [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+    this._hands = hands;
+    this.addProperties("_legs", "_hands");
+  }
+
+  get saying() {
+    return this._phrase;
+  }
+}
+
+class Animal extends Creature {
```
Comment (OleksiyRudenko) [inline submissions/kotlyar-andrey/oop_exercise/index.js:64]: Strictly speaking humans are animals as well. We need to resolve the matter in a consistent way.

**Review by OleksiyRudenko — CHANGES_REQUESTED** [review]
@kotlyar-andrey very well done!
OOP is not really good for building hybrids. And this is one of the learnings from this exercise to take. The proper approach would be composition. You do not need to fix anything in these regards. Just reflect on possible flaws of OOP-like approach.

Still we need to add some consistency with regards to the business domain. Please check comments below.

**Comment by kotlyar-andrey** [comment]
Thanks for the review!
I've made changes, please re-review.

File: submissions/kotlyar-andrey/oop_exercise/index.js [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+        .concat(`; ${friends}`)
+    );
+  }
+}
+
+class Animal extends Creature {
+  constructor(species, name, gender, phrase, legs, hands) {
+    super(species, name, gender, phrase);
+    this._legs = legs;
+    this._hands = hands;
```
Comment (OleksiyRudenko) [inline submissions/kotlyar-andrey/oop_exercise/index.js:55]: Do all animals have hands?
Imagine we popultae the world with birds, fish, rhinos and deer.
OOP helps to model the real world quite accurately avoiding defining properties and behavior that the counterparts in the real world do not have.

**Review by OleksiyRudenko — CHANGES_REQUESTED** [review]
@kotlyar-andrey good fix. Yet we need to make this code compliant with OOP/SOLID principles.
You will find extra materials [here](https://github.com/OleksiyRudenko/a-tiny-JS-world#leveling-your-skills-up) helpful.
Feel free asking specific questions based on your learnings.

**Comment by kotlyar-andrey** [comment]
I think, I understand what you mean. Review, please)

File: submissions/kotlyar-andrey/oop_exercise/index.js [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+  }
+
+  addProperties(...newProperties) {
+    this._properties = [...this._properties, ...newProperties];
+  }
+
+  display() {
+    const friends = this.getFriends();
+    const saying = this.getSaying();
+    print(
```
Comment (OleksiyRudenko) [inline submissions/kotlyar-andrey/oop_exercise/index.js:46]: A class shouldn't know anything about external environment.
Using a side effect (`print` is a side effect) also makes the class less flexible.

File: submissions/kotlyar-andrey/oop_exercise/index.js [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+      this._properties
+        .map(prop => this[prop])
+        .join("; ")
+        .concat(`; ${saying}; ${friends}`)
```
Comment (OleksiyRudenko) [inline submissions/kotlyar-andrey/oop_exercise/index.js:50]: Not a really good approach to have a list of non-array props and handle array props in a totally different way. Besides, this method does two things: converts array props to strings, builds object presentation string.

What if we have 20 more array props? What if we have props that keep objects?

We need a more consistent approach.

Suggest an approach before implementing changes in these regards.

**Review by OleksiyRudenko — CHANGES_REQUESTED** [review]
@kotlyar-andrey very good.
Although there are two flaws that will make the project harder to maintain and scale if we want to expand it.

File: submissions/kotlyar-andrey/oop_exercise/index.js (hunk 4 above)
Comment (kotlyar-andrey) [reply submissions/kotlyar-andrey/oop_exercise/index.js:46]: Ok, I can rename display to getInfo and return value
  ```
getInfo() {
    const friends = this.getFriends();
    const saying = this.getSaying();
    return this._properties
      .map(prop => this[prop])
      .join("; ")
      .concat(`; ${saying}; ${friends}`);
  }
```
  End then print inhabitants
 ```
 inhabitants.forEach(creature => {
    print(creature.getInfo());
});
```

Comment (OleksiyRudenko) [reply submissions/kotlyar-andrey/oop_exercise/index.js:46]: Making class method just return some data is definitely a good design decision.

File: submissions/kotlyar-andrey/oop_exercise/index.js (hunk 5 above)
Comment (kotlyar-andrey) [reply submissions/kotlyar-andrey/oop_exercise/index.js:50]: > Not a really good approach to have a list of non-array props and handle array props in a totally different way. Besides, this method does two three things: converts array props to strings, builds object presentation string, prints the presentation string.

We can refactor this method to three another function

> What if we have 20 more array props? What if we have props that keep objects?

Every class has addProperties method for adding new properties. 
If we want to get another data in child classes, we can override method getAnotherData

 ```
getDataProperties() {
    return this._properties.map(prop => this[prop]).join("; ");
  }

  getAnotherData() {
    const friends = this.getFriends();
    const saying = this.getSaying();
    return `; ${friends}; ${saying}`;
  }

  getInfo() {
    return this.getDataProperties() + this.getAnotherData();
  }
```

This is all what I can offer

File: submissions/kotlyar-andrey/a-tiny-JS-world/index.js [hunk 6]
```diff
@@ -5,72 +5,132 @@
...
+
+/**
+ * Общий класс для всех животных с ногами/лапами
+ */
+class Animal extends Creature {
+  constructor(species, name, gender, phrase, legs) {
+    super(species, name, gender, phrase);
+    this._legs = legs;
+    this.addProperties("_legs");
+  }
```
Comment (OleksiyRudenko) [inline submissions/kotlyar-andrey/a-tiny-JS-world/index.js:66]: If some class (e.g. `Creature`) is extended by one and only one class (e.g. `Animal`) this means they can be effectively merged. Yes, we may have legless creatures in future, but that's exactly the point when we would split a base class and/or add yet another base class on the top of hierarchy tree.

File: submissions/kotlyar-andrey/a-tiny-JS-world/index.js [hunk 7]
```diff
@@ -5,72 +5,132 @@
...
+    this._wings = 2;
+    this.addProperties("_wings");
+  }
+}
+
+class Human extends Animal {
+  constructor(name, gender, phrase) {
+    super("human", name, gender, phrase, 2);
+    this._hands = 2;
+    this.addProperties("_hands");
```
Comment (OleksiyRudenko) [inline submissions/kotlyar-andrey/a-tiny-JS-world/index.js:84]: We delegate child class property handling to a base class.
This is a legal approach. Although we should always consider first that every class handles its own and only its own properties. This way the handling logic will be encapsulated within the owning class and not spread across the hierarchy.

**Review by OleksiyRudenko — APPROVED** [review]
@kotlyar-andrey great job done!
Takes notes of the comments below.

**Pull request merged**